### PR TITLE
[DeepSeek] Add all_to_all_single impl; Enable backward

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #956
* __->__ #989
* #958
* #954
* #952
* #941

Adding `dist.all_to_all_single` as token shuffle method. 
This is made the default option for better support of backward pass.

The alternative shuffle method --`on_device_all_to_all_v` -- will be enabled only when user calls
`model.setup_symm_mem()`
and is limited to inference only.

### CODE REVIEW

File: torchtitan/experiments/deepseek_v3/model.py [hunk 1]
```diff
@@ -570,23 +576,42 @@ def moe_forward(self, x, topk_ids, topk_weight):
...
+                MoE.input_splits,
+                self.ep_group,
+            )
+            with torch.no_grad():
+                # Received tokens from all other ranks. TODO: use mask instead
+                received = MoE.output_splits.sum()
+            # TODO: don't use `received`
+            gathered_tokens = MoE.token_gather_buf[:received]
+        else:  # "torch_all_to_all"
+            # Prepare input ans output splits
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:603]: Typo: and

also, the point of this else branch is just to offer  EP>8 support for now, right?  and perhaps also benchmark between the two on EP=8

File: torchtitan/experiments/deepseek_v3/model.py [hunk 2]
```diff
@@ -1189,3 +1229,11 @@ def _reorder_cache(past_key_values, beam_idx):
                 ),
             )
         return reordered_past
+
+    # Setup Symmetric Memory for MoE token shuffle.
+    # Supports inference currently.
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:1234]: doesn't support training you mean?
there was a comment on the first PR in the stack that points out a bug in backward formula for all2all.  Otoh I thought training was supported.

Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/model.py:1234]: Yeah, I put a comment in the code:
```
For training, this is disabled for now. Reason is that autograd requires tensors not be modified a second time, this conflicts with our wish of sharing the symm mem across layers and/or PP microbatches.
```
Till we find a workaround, this is disabled for backward.

Comment (EugenHotaj) [reply torchtitan/experiments/deepseek_v3/model.py:1234]: We noticed this as well when debugging the backward pass / the OOM mentioned in #979. 

One way we were able to get around both issues is by using the symmetric memories purely as buffers and passing them directly to `on_device_all_to_all_v` and calling e.g. `token_gather_buffer.copy_(tensor)` inside the `OnDeviceAllToAllV.forward` function and changing `OnDeviceAllToAllV.backward` accordingly. 

But tbh I'm not sure this resulted in correct gradients...

**Comment by EugenHotaj** [comment]
Noob q: do we need to allreduce the gradients for the non-MoE parameters or is this automatically handled somehow?

File: torchtitan/experiments/deepseek_v3/model.py (hunk 2 above)
Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/model.py:1234]: @EugenHotaj Yeah, we may need to go the way you said.

**Comment by kwen2501** [comment]
Great great great question!
In PR #956 we apply DP differently -- for non-shared experts, we apply FSDP; for all other parts (including shared experts), we apply HSDP. Meaning, for the "all other parts", they are replicates in the EP dimension, hence they need DDP in that dimension. DDP + FSDP = HSDP.

**Comment by EugenHotaj** [comment]
@kwen2501 if I understand you correctly, the HSDP wrapping in #956 will automatically handle allreducing the grads for the non-EP parts?

**Comment by kwen2501** [comment]
Yep.

File: torchtitan/experiments/deepseek_v3/model.py [hunk 3]
```diff
@@ -600,34 +625,49 @@ def moe_forward(self, x, topk_ids, topk_weight):
...
+        # Prepare buffer for tokens processed by experts
+        if self.shuffle_method == "symm_mem":
+            # Take necessary space from `token_send_buf` symm mem because we are
+            # going to send them out after expert processing
+            processed_tokens = MoE.token_send_buf[: gathered_tokens.shape[0]]
+        else:  # "torch_all_to_all"
+            processed_tokens = torch.empty_like(gathered_tokens)
+
+        # This part processes the tokens routed to the local experts.
+        # TODO: can we use group GEMM here?
```
Comment (lessw2020) [inline torchtitan/experiments/deepseek_v3/model.py:637]: yes,  will integrate my MG group GEMM here. (https://github.com/pytorch/torchtitan/pull/967)

File: torchtitan/experiments/deepseek_v3/model.py [hunk 4]
```diff
@@ -600,34 +625,49 @@ def moe_forward(self, x, topk_ids, topk_weight):
...
+                processed_tokens,
+                input_splits.tolist(),
+                output_splits.tolist(),
+                self.ep_group,
+            )
 
-        returned_tokens = torch.empty_like(gathered_tokens)
-        returned_tokens[idxs] = gathered_tokens
+        output_tokens = torch.empty_like(returned_tokens)
+        output_tokens[idxs] = returned_tokens
```
Comment (lessw2020) [inline torchtitan/experiments/deepseek_v3/model.py:664]: approx, in here we can add an un-permute kernel to handle and speedup the returned token re-ordering.

**Review by lessw2020 — APPROVED** [review]
lgtm. 
Stamping to get this landed!

**Comment by kwen2501** [comment]
Closing via https://github.com/pytorch/torchtitan/pull/1002
